### PR TITLE
API for bans, abstracted to support quiets/exempt

### PR DESCRIPTION
**Pull request by mbax**
Done:
* Create `Ban` element
* Create a list of `Ban`s when `MODE #channel +b` gets a reply

TODO:
* [x] Track bans for channels the client is in
* [x] Convenience method for removing a ban
* [x] Event for ban list being updated (e.g. shortly after channel join)
* [x] *Maybe* toggle tracking of bans per channel.

### CODE REVIEW

**Comment by Zarthus** [comment]
If you're going to do

>Maybe toggle tracking of bans per channel

You might as well make it configurable:

```
                                // v    v only gettable by ops i think
Char[] modesToTrack = { 'b', 'q', 'I', 'e' }
```

Also just a global toggle would be nice.

File: src/main/java/org/kitteh/irc/client/library/element/mode/Ban.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    Channel getChannel();
+
+    @Nonnull
+    Client getClient();
+
+    @Nonnull
+    String getId();
+
+    @Nonnull
+    Optional<Integer> getTimeLeft();
```
Comment (Zarthus) [inline src/main/java/org/kitteh/irc/client/library/element/mode/Ban.java]: It's not the time left, but the date the ban was set (as best known as by the IRCd. Which might not be very reliable to begin with)

Comment (kashike) [inline src/main/java/org/kitteh/irc/client/library/element/mode/Ban.java]: This could be an `Instant` (going off what @Zarthus noted above).

Comment (bendem) [inline src/main/java/org/kitteh/irc/client/library/element/mode/Ban.java]: Or `Duration` if @Zarthus was wrong (you never know :innocent:)

Comment (Zarthus) [inline src/main/java/org/kitteh/irc/client/library/element/mode/Ban.java]: It's an unix timestamp, for the curious.

File: src/main/java/org/kitteh/irc/client/library/implementation/ModeData.java [hunk 2]
```diff
@@ -142,20 +143,20 @@ public Client getClient() {
...
-            return this.timeLeft;
+        public Optional<Instant> getCreationTime() {
+            return this.creationTime;
         }
 
         @Nonnull
         @Override
         public String toString() {
-            return new ToStringer(this).add("client", this.client).add("channel", this.channel).add("id", this.id).add("banner", this.banner).add("timeleft", this.timeLeft).toString();
+            return new ToStringer(this).add("client", this.client).add("channel", this.channel).add("id", this.id).add("banner", this.banner).add("timeleft", this.creationTime).toString();
```
Comment (kashike) [inline src/main/java/org/kitteh/irc/client/library/implementation/ModeData.java]: `"timeLeft"` -> `"creationTime"`

File: src/main/java/org/kitteh/irc/client/library/event/channel/ChannelModeInfoListEvent.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import org.kitteh.irc.client.library.event.abstractbase.ChannelEventBase;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A list of mode info is available!
+ */
+public class ChannelModeInfoListEvent extends ChannelEventBase {
```
Comment (kashike) [inline src/main/java/org/kitteh/irc/client/library/event/channel/ChannelModeInfoListEvent.java]: I assume this is for bans, quiets, etc.
There's currently no way, that I can see, to determine which type of list this is.

Comment (mbax) [inline src/main/java/org/kitteh/irc/client/library/event/channel/ChannelModeInfoListEvent.java]: From latest commit - "Needs way of knowing it's a ban"

File: src/main/java/org/kitteh/irc/client/library/implementation/ActorProvider.java [hunk 4]
```diff
@@ -209,6 +214,35 @@ IRCChannelSnapshot snapshot() {
...
+            }
+            this.markStale();
+        }
+
+        void trackModeInfo(boolean add, @Nonnull ModeInfo modeInfo) {
+            if (add) {
+                this.modeInfoLists.get(modeInfo.getMode().getChar()).add(modeInfo);
+            } else {
+                Iterator<ModeInfo> iterator = this.modeInfoLists.get(modeInfo.getMode().getChar()).iterator();
+                while (iterator.hasNext()){
```
Comment (kashike) [inline src/main/java/org/kitteh/irc/client/library/implementation/ActorProvider.java]: `) {`

File: src/main/java/org/kitteh/irc/client/library/util/Mask.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+package org.kitteh.irc.client.library.util;
+
+import org.kitteh.irc.client.library.element.User;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Represents a mask that can match a {@link User}.
+ */
+public class Mask {
```
Comment (Zarthus) [inline src/main/java/org/kitteh/irc/client/library/util/Mask.java]: `existsIn(List<ModeInfo>)` utility? (or for `User`)

**Comment by Zarthus** [comment]
Sidecomments:

Some IRCds support `/mode #channel bqIe`, does KICL support that? (I have no idea where this feature is documented, or how to detect it) -- sounds like a neat optimization.

**Comment by mbax** [comment]
KICL does not currently handle rate limiting. As for querying modes that way, no it doesn't either.

File: src/main/java/org/kitteh/irc/client/library/util/Mask.java (hunk 5 above)
Comment (mbax) [inline src/main/java/org/kitteh/irc/client/library/util/Mask.java]: See https://github.com/KittehOrg/KittehIRCClientLib/issues/130

**Pull request merged**